### PR TITLE
Scanner

### PR DESCRIPTION
**Pull request by jared-nishikawa**
Some base code hacked up that we can play around with.

The scanner tokenizes input into SGF-specific tokens.
For example: `(;GM[1]FF[4]CA[UTF-8]AP[CGoban:3]ST[2]` becomes
`["(", ";", "GM", "[", "1", "]", "FF", "[", "4", "]", "CA", "[", "UTF-8", "]", "AP", "[", "CGoban:3", "]", "ST", "[", "2", "]"]`

The parser takes these tokens and parses them into nodes and branches. Nodes start with a semicolon ";" and branches start with a left paren "(".  Currently the parser just stores all these nodes in a map, indexed by an int.  Eventually, we will want to update this data structure to retain the tree structure of the game variations.

### CODE REVIEW

File: sgf/parser.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
+package sgf
+
+import (
+	"errors"
+	"io"
+)
+
+// A node is comprised of key-value pairs (might hold other info someday)
+type Node struct {
+	Fields map[string]string
```
Comment (artasparks) [inline sgf/parser.go]: Typically, I've called these Properties.

Should be a list of string since SGF properties can be repeated.

Also, I would expect a Node to be a recursive type:

Children []*Node

File: sgf/parser.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
+package sgf
+
+import (
+	"errors"
+	"io"
+)
+
+// A node is comprised of key-value pairs (might hold other info someday)
+type Node struct {
```
Comment (artasparks) [inline sgf/parser.go]: Let's move to game/

File: sgf/parser.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+func NewNode() *Node {
+	n := &Node{}
+	n.Fields = make(map[string]string)
+	return n
+}
+
+// A game record (sgf) struct for now is just a collection of nodes
+// TODO: this struct should retain the branch structure of game variations
+type Sgf struct {
```
Comment (artasparks) [inline sgf/parser.go]: Let's move this type to game/game.go (and call it Game)

File: sgf/parser.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+	Index int
+	Nodes map[int]*Node
```
Comment (artasparks) [inline sgf/parser.go]: Typically, I would expect a Game to have a single Root node.

File: sgf/parser.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
+package sgf
```
Comment (artasparks) [inline sgf/parser.go]: Probably package for parser should be parse or parser.

**Review by artasparks — COMMENTED** [review]
Couple of initial comments

File: sgf/scanner.go [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+	Raw  string
+}
+
+// Scanner wraps bufio.Reader but has the public Scan() function to scan tokens
+type Scanner struct {
+	*bufio.Reader
+}
+
+// Creates a new scanner object out of an io.Reader interface
+func NewScanner(r io.Reader) *Scanner {
```
Comment (artasparks) [inline sgf/scanner.go]: It's probably helpful to have a method 

FromString(s string) {

}

Or honestly, just a Parse(s string) method is probably more helpful.

File: sgf/scanner.go [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+		return &Token{
+			Type: RightBracket,
+			Raw:  "]",
+		}
+	case ';':
+		return &Token{
+			Type: Semicolon,
+			Raw:  ";",
+		}
+	default:
```
Comment (artasparks) [inline sgf/scanner.go]: Actually, wouldn't this be an error/unexpected token?

**Review by artasparks — COMMENTED** [review]
Might be helpful: https://github.com/Kashomon/bodoko/blob/master/core/sgf_parser.go. A previous parser I wrote in go.

File: sgf/parser.go (hunk 1 above)
Comment (jared-nishikawa) [reply sgf/parser.go]: Sounds good, I've redone Node to have both children and parent (to traverse either direction, if needed)

File: sgf/parser.go (hunk 2 above)
Comment (jared-nishikawa) [reply sgf/parser.go]: Done

File: sgf/parser.go (hunk 3 above)
Comment (jared-nishikawa) [reply sgf/parser.go]: Also moved everything to

internal/
-  game/
-  parser/
-  scanner/

File: sgf/parser.go (hunk 4 above)
Comment (jared-nishikawa) [reply sgf/parser.go]: Agreed, fixed

File: sgf/parser.go (hunk 5 above)
Comment (jared-nishikawa) [reply sgf/parser.go]: I made an attempt at cleaner packaging by making scanner, parser, and game directories inside of "internal"

File: sgf/scanner.go (hunk 6 above)
Comment (jared-nishikawa) [reply sgf/scanner.go]: Hmmmm... I put in FromString and FromReader for flexibility.  Are you saying we should scrap FromReader?

File: sgf/scanner.go (hunk 7 above)
Comment (jared-nishikawa) [reply sgf/scanner.go]: Agreed, fixed

File: sgf/parser.go (hunk 1 above)
Comment (jared-nishikawa) [reply sgf/parser.go]: I also didn't realize the properties can be repeated.  Changed to `map[string][]string`

File: internal/game/game.go [hunk 8]
```diff
@@ -0,0 +1,27 @@
+package game
```
Comment (artasparks) [inline internal/game/game.go]: I don't think Internal is right for game. I think the goal should be to provide generally useful go libraries -- and certainly Game would be one of those. So, I'd make it a publicly available library at the top level.

File: internal/parser/parser.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return g, nil
+}
+
+// parse, for example "PW[Player White]" into
+// field="PW"
+// value="Player White"
+func (p *Parser) parseFieldValue() (string, string, error) {
```
Comment (artasparks) [inline internal/parser/parser.go]: How does this handle something like AW[pa][qa][nb][ob]?

File: internal/parser/parser.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+// scanSkipWhitespace is a wrapper around scan() that skips whitespace
+func (p *Parser) scanSkipWhitespace() *scanner.Token {
+	tok := p.scan()
+	if tok.Type == scanner.Whitespace {
+		tok = p.scan()
+	}
+	return tok
+}
+
+// Parses the root branch
```
Comment (artasparks) [inline internal/parser/parser.go]: nit: Parse have form "Parse parses the root branch"

File: internal/parser/parser.go [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+// field="PW"
+// value="Player White"
+func (p *Parser) parseFieldValue() (string, string, error) {
+	var tok *scanner.Token
+	var field string
+	var value string
+	// parse a string
+	// TODO: better erroring
+	if tok = p.scanSkipWhitespace(); tok.Type != scanner.String {
+		return "", "", errors.New("Corrupted sgf: 3")
```
Comment (artasparks) [inline internal/parser/parser.go]: what does corrupted sgf: 3 mean?

File: internal/scanner/scanner.go [hunk 12]
```diff
@@ -0,0 +1,166 @@
+package scanner
```
Comment (artasparks) [inline internal/scanner/scanner.go]: Generally packages should have the form

// Package scanner does ...

File: internal/scanner/scanner.go [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+	RightBracket
+	Semicolon
+	Backslash
+	String
+	Illegal
+)
+
+var eof = rune(0)
+
+// isWhiteSpace is a check to see if a rune is whitespace (space, tab, newline, carriage return)
```
Comment (artasparks) [inline internal/scanner/scanner.go]: nit: wrap comments at 80 chars for readability.

File: internal/scanner/scanner.go [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+	return ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r'
+}
+
+// isSpecial is a check for special characters used in SGF files
+func isSpecial(ch rune) bool {
+	return ch == '(' || ch == ')' || ch == '[' || ch == ']' || ch == ';'
+}
+
+// The scanner will "tokenize" the input into Token structs
+type Token struct {
```
Comment (artasparks) [inline internal/scanner/scanner.go]: Nit: Comment should have form "Token does blah"

Golint should catch these sorts of errors

**Review by artasparks — COMMENTED** [review]
Still haven't done a full review XD. Couple more comments. I'd recommend running Golint to catch any small errors -- I'll eventually make some presubmits to catch that.

File: internal/parser/parser.go [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+		return "", "", errors.New("Corrupted sgf: 4")
+	}
+
+	// parse anything until a right bracket
+	for {
+		tok = p.scan()
+		if tok.Type == scanner.Eof {
+			return "", "", errors.New("EOF")
+		}
+		if tok.Type == scanner.RightBracket {
```
Comment (artasparks) [inline internal/parser/parser.go]: What happens if you have a right-bracket that's part of a comment? 

[\\]] Should be valid. 

https://github.com/Kashomon/glift-core/blob/master/src/parse/sgf_parsing_test.js has a bunch of edge cases I've collected over the years.

**Comment by jared-nishikawa** [comment]
TIL about golint :)

File: internal/parser/parser.go (hunk 15 above)
Comment (jared-nishikawa) [reply internal/parser/parser.go]: The parser handles `\]`
I'll have a look through the other edge cases

File: internal/parser/parser.go (hunk 9 above)
Comment (jared-nishikawa) [reply internal/parser/parser.go]: Not correctly, I'll tell you that :)

File: internal/parser/parser.go (hunk 11 above)
Comment (jared-nishikawa) [reply internal/parser/parser.go]: If it didn't parse a string, throw an error.  The number is meaningless, just for testing

File: internal/game/game.go (hunk 8 above)
Comment (jared-nishikawa) [reply internal/game/game.go]: How about the parser?

Comment (artasparks) [reply internal/game/game.go]: I'd keep parser public too -- seems like a generally useful tool. Scanner might deserve to be internal, but that could just be private variables within parser

Generally, I feel internal directories should be reserved as a last resort when your private types have grown too unwieldy and you want to share them with other library.

Comment (jared-nishikawa) [reply internal/game/game.go]: Thanks for clarifying, this is super useful for other projects I have at work

File: internal/parser/parser.go (hunk 9 above)
Comment (jared-nishikawa) [reply internal/parser/parser.go]: Just had a shot at a fix

File: game/game.go [hunk 16]
```diff
@@ -0,0 +1,28 @@
+package game
+
+// Game contains the game tree information
+type Game struct {
+	Root *Node
+}
+
+// NewGame creates a Game struct
+func NewGame() *Game {
```
Comment (artasparks) [inline game/game.go]: Nit: because the package is Game you can just say New(), because it'll be called as `game.New()`

File: scanner/scanner.go [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+)
+
+// constants
+const (
+	EOF = iota
```
Comment (artasparks) [inline scanner/scanner.go:11]: nit: You could use a string Consts for these instead of ints.

You also might provide a custome type here for clarity.

type ControlChar int

or similar

Or maybe

type ControlChar rune

File: scanner/scanner.go [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+	return ch
+}
+
+// Go back one rune
+func (s *Scanner) unread() {
+	s.UnreadRune()
+}
+
+// Scan is the function that tokenizes the input and outputs tokens
+func (s *Scanner) Scan() *Token {
```
Comment (artasparks) [inline scanner/scanner.go:67]: nit: I'd recommend passing back the number of chars read so the caller can can keep track of of the column / line # for error messaging. Or maybe # of chars and # of newlines.

File: scanner/scanner.go [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+		}
+	default:
+		return &Token{
+			Type: Illegal,
+			Raw:  "",
+		}
+	}
+}
+
+// scanWhiteSpace consumes as much as whitespace as possible
```
Comment (artasparks) [inline scanner/scanner.go:123]: nit: s/scanWhiteSpace/scanWhitespace

File: scanner/scanner.go [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+var eof = rune(0)
+
+// isWhiteSpace is a check to see if a rune is whitespace
+// (space, tab, newline, carriage return)
+func isWhitespace(ch rune) bool {
+	return ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r'
+}
+
+// isSpecial is a check for special characters used in SGF files
+func isSpecial(ch rune) bool {
```
Comment (artasparks) [inline scanner/scanner.go:32]: nit: I might call these ControlChars -- isControlChar

File: parser/parser_test.go [hunk 21]
```diff
@@ -0,0 +1,59 @@
+package parser_test
+
+import (
+	"github.com/otrego/clamshell/parser"
+	"testing"
+)
+
+func TestCreate(t *testing.T) {
```
Comment (artasparks) [inline parser/parser_test.go:8]: I'd recommend more extensive tests here especially smaller examples that test specific features. TableTests are a nice way to set this up: https://github.com/golang/go/wiki/TableDrivenTests

**Review by artasparks — COMMENTED** [review]
Bunch of nits: only thing that I think we should improve before submitting is adding a bit more extensive testing.

**Comment by artasparks** [comment]
I'd like to merge this in so we can start to iterate. Do you have much more you're planning on adding?

**Comment by jared-nishikawa** [comment]
@Kashomon I haven't made all of the changes you've asked for:

- Better testing
- Handling of edge cases
- Custom type for control chars
- Scan() needs to return number of characters scanned

I was thinking I would need at least another week to address all of the edge cases and write better tests.

Option 1: We could merge it in while not in a fully functional state.
Option 2: If your bodoko/sgf_parser is functional, we could just use that.

**Comment by artasparks** [comment]
Let's merge as is and I'll file some issue to follow-up. =)
